### DISCUSSION TITLE
Anyone using CC-U03 Thermostats?

### DISCUSSION THREAD

**Discussion by 0n3man** [ccutrer/waterfurnace_aurora#74] [Q&A] [answered]
I been running ccutrer's software to integrate with HA for a while now without issue.  However my CC-U02 Thermostats generates heat, thus causing the thermostat to think the temperature is about 2 degrees hotter than reality.  I'd be curious if anyone has had this issue?  Waterfurance no longer shows the CC-U02 on their site and have replaced it with the CC-U03 model.  I'm curious if anyone is running ccutrer's software in combination with the CC-U03, and if yes does it work without issue?

**Comment by nyneaxis** [comment]
I am pretty sure I am running really old models. I have four zones. None of the stats looks anything like the ones on the website. I just had my unit installed this summer. Is there a way to purchase newer ones and install them? 

@0n3man if you have Discord message me please, username emetrixlabs. Would like to pick your brain on a few things relate to WF and the ccutrer software.

**Comment by KO-RAM** [comment, marked as answer]
Not sure if you're still interested, but I have a CC-U03 (model: TPCC32U03) thermostat on my 7-series system. I have not had any issues with it. Temperatures stay within 0.5 degree F of set point. I recently set up ccutrer's software and everything is running great.

**Reply by 0n3man** [reply]
Thanks for the update.  I haven't yet updated my thermostat, but will probably do so shortly.

**Reply by nyneaxis** [reply]
@KO-RAM where can I purchase the CC-U03 model? I have one main stat and three zone stats. Not sure why the installer didn't put in the newest model last year but they are the original CC-U01.

**Comment by 0n3man** [comment]
Any WaterFurnace dealer in your area.  WF will only sell through an HVAC
company.  The thing is expensive.  I saw something online indicating it
should sell for around $750, but my dealer wants around $1150.

**Reply by nyneaxis** [reply]
WHAT!!!! That is insane. What stat should cost that much????

**Reply by KO-RAM** [reply]
Yeah, they are expensive. I had to replace my 5 series system in 2021 and so I put in a 7 series with all the bells and whistles (performance, refrigeration, and energy monitoring options). The $750 sounds about right for the thermostat. Keep in mind that the thermostat is not your average Nest type furnace thermostat. The WF thermostat communicates on Modbus with the ABC and AXB Aurora control boards and gives you many more capabilities.

**Reply by ccutrer** [reply]
> The WF thermostat communicates on Modbus with the ABC and AXB Aurora control boards

To clarify, the thermostat bus is RS-485, but it's _not_ ModBus.

But yes, it is a communicating thermostat, which 99% of thermostats are not.

Still doesn't justify the insane prices. Touchscreens are cheap components nowadays. RS-485 is multiple orders of magnitude less complex than the WiFi and IP stacks in a Nest thermostat. Not that any one company develops any of those anymore. Basically all the embedded devices get an SDK from the MCU OEM, which includes all the network connectivity in the first place (usually based on LWIP for the IP side; the WiFi side might still be proprietary). I'm sure part of it is much lower volumes, so their costs are higher. But part of it is definitely that it IS a communicating thermostat with a proprietary protocol, which then means they have no competition, nor do consumers like us have much ability to even shop around, so they charge what they want.

Anyhow, I had to replace a thermostat a couple years ago because I accidentally cracked a screen. I got lucky to find one on eBay for like $250. Good luck on your search.

**Reply by KO-RAM** [reply]
I stand corrected. I had assumed since it was RS485 they were running the Modbus protocol to the thermostat also. I was going to try to get the outdoor temperature off the thermostat thinking it was Modbus, but that may not work the way I thought. Have you tried to pull the temperature in? I thought I saw a reference in the source code to it somewhere. 

I agree with the comments on the cost of these thermostats. WF is definitely making money on these.

**Reply by 0n3man** [reply]
I've asked for an estimate on the TPCM32U04A as its listed as being a communicating thermostat without a touch screen.  It's also listed to work with the WF symphony capability.  As such I expect it should work with this software.  I was hoping someone on the list was using this non touch screen thermostat already.   My expectation is it will not be that much cheaper.  I'll post the estimate once I've received it.

**Reply by 0n3man** [reply]
[ccutrer](https://github.com/ccutrer) do you happen to know if I could get a smart thermostat that works with HA installed, and still utilize your code to pull all the furnace data?

**Reply by ccutrer** [reply]
> Have you tried to pull the temperature in?

The outdoor temperature (if available) is on the ABC-side ModBus bus, so yes, it's exposed by this gem.

> ccutrer do you happen to know if I could get a smart thermostat that works with HA installed, and still utilize your code to pull all the furnace data?  

This gem will work regardless of what thermostat you have controlling the heat pump. In theory you can use a non-communicating thermostat with the WaterFurnace, and the gem will still work. A first party thermostat may be required if you have IZ2, though - I'm not certain. Maybe only the MasterStat? (The MasterStat is the only way to configure relative zone sizes and other staging settings in my IZ2 system). Though if you don't have a communicating thermostat you can't set its setpoint(s) via Aurora or this gem. Probably not an issue, since your goal is to use some other thermostat that's directly compatible with other automation systems.

**Reply by KO-RAM** [reply]
I haven't tried to pull in the outdoor temperature yet.  Is that done through the aurora_fetch command? I tried the example you have, but I'm getting an error:  

kevin@wfmodbuspi:~ $ aurora_fetch /dev/ttyHeatPunp 745-746
/var/lib/gems/3.1.0/gems/ccutrer-serialport-1.1.0/lib/ccutrer/serial_port.rb:7:in `initialize': No such file or directory @ rb_sysopen - /dev/ttyHeatPunp (Errno::ENOENT)
        from /var/lib/gems/3.1.0/gems/ccutrer-serialport-1.1.0/lib/ccutrer/serial_port.rb:7:in `initialize'
        from /var/lib/gems/3.1.0/gems/waterfurnace_aurora-1.5.5/lib/aurora/abc_client.rb:41:in `new'
        from /var/lib/gems/3.1.0/gems/waterfurnace_aurora-1.5.5/lib/aurora/abc_client.rb:41:in `open_modbus_slave'
        from /var/lib/gems/3.1.0/gems/waterfurnace_aurora-1.5.5/exe/aurora_fetch:54:in `<top (required)>'
        from /usr/local/bin/aurora_fetch:25:in `load'
        from /usr/local/bin/aurora_fetch:25:in `<main>'
kevin@wfmodbuspi:~

**Reply by ccutrer** [reply]
Outdoor temp is in register 31,002 if you're using aurora_fetch. Your error is about it not being able to find your serial port, though. `ttyHeatPunp` is almost certainly not the name of your device. And it would only be `ttyHeatPump` if you've set up udev rules to alias it. If you have no other USB serial devices on your system, it's likely `ttyUSB0`.

**Reply by KO-RAM** [reply]
It looks like the outdoor temperature isn't getting picked up on the ABC register. It's returning 0.0 degrees F on address 31003. Current outdoor temperature on the thermostat is 8 deg F. Here's the response (I fetched some of the registers around the outdoor temperature initially). 
kevin@wfmodbuspi:~ $ aurora_fetch /dev/ttyUSB0 31000-31008
??? (31000): 0 (0x0000)
??? (31001): 0 (0x0000)
??? (31002): 0 (0x0000)
Outdoor Temp (31003): 0.0°F
??? (31004): 0 (0x0000)
IZ2 Demand (31005): {:fan_demand=>0, :unit_demand=>0}
??? (31006): 0 (0x0000)
Zone 1 Ambient Temperature (31007): 0.0°F
Zone 1 Configuration 1 (31008): {:fan=>:auto, :on_time=>0, :off_time=>5, :cooling_target_temperature=>36, :heating_target_temperature_carry=>0}
kevin@wfmodbuspi:~ $ aurora_fetch /dev/ttyUSB0 31003
Outdoor Temp (31003): 0.0°F
kevin@wfmodbuspi:~ $ 

Any thoughts on what might be going on?

**Reply by ccutrer** [reply]
I suspect the register I'm using is an IZ2-only register (which it looks like you don't have). The proper "regular" register will need to be discovered.

Possible candidates that match the current temperature on my system: 568, 742, 2011, 2118. Do any of those registers have a non-zero value for you?

**Reply by KO-RAM** [reply]
You are correct, only one zone on my 7-series system. Register 742 looks like a possibility if the value is in tenths of a degree.  The thermostat outdoor temperature is 14 deg F. The value of register 742 is 146.  If it was tenths it could be 14.6 deg F. I can see what the register shows as the outdoor temperature changes and see if my theory is correct. Over about an hour period it dropped from 146 to 140 while the thermostat outdoor temperature stayed at 14 deg F. The other ranges of registers are all zeros.  Here's what the fetch provides:  

kevin@wfmodbuspi:~ $ aurora_fetch /dev/ttyUSB0 568
??? (568): 0 (0x0000)
kevin@wfmodbuspi:~ $ aurora_fetch /dev/ttyUSB0 742
??? (742): 146 (0x0092)
kevin@wfmodbuspi:~ $ aurora_fetch /dev/ttyUSB0 2011
??? (2011): 0 (0x0000)
kevin@wfmodbuspi:~ $ aurora_fetch /dev/ttyUSB0 2118
??? (2118): 0 (0x0000)
kevin@wfmodbuspi:~ $

Here are some of the registers around register 742:  
kevin@wfmodbuspi:~ $ aurora_fetch /dev/ttyUSB0 720-750
??? (720): 0 (0x0000)
??? (721): 0 (0x0000)
??? (722): 0 (0x0000)
??? (723): 0 (0x0000)
??? (724): 0 (0x0000)
??? (725): 0 (0x0000)
??? (726): 0 (0x0000)
??? (727): 0 (0x0000)
??? (728): 0 (0x0000)
??? (729): 0 (0x0000)
??? (730): 0 (0x0000)
??? (731): 0 (0x0000)
??? (732): 0 (0x0000)
??? (733): 0 (0x0000)
??? (734): 0 (0x0000)
??? (735): 0 (0x0000)
??? (736): 0 (0x0000)
??? (737): 0 (0x0000)
??? (738): 0 (0x0000)
??? (739): 0 (0x0000)
Entering Air (740): 68.5°F
Relative Humidity (741): 41%
??? (742): 138 (0x008a)
??? (743): 3 (0x0003)
??? (744): 81 (0x0051)
Heating Set Point (745): 68.0°F
Cooling Set Point (746): 73.0°F
Ambient Temperature (747): 68.0°F
??? (748): 50 (0x0032)
??? (749): 45 (0x002d)
??? (750): 0 (0x0000)
kevin@wfmodbuspi:~ $

**Reply by ccutrer** [reply]
Yes, most of the temperature registers are in tenths of degrees Fahrenheit. Interesting that you're seeing sub-integer temperatures - I only get integer ones. I wonder if it's because the IZ2 doesn't support sub-integer sensing, or is just not passing through sub-integer. But now that I think about it more, I recall that when connecting the exterior temperature sensor, I had the option of connecting it to the thermostat, or to the IZ2 controller, whichever is more convenient. I chose IZ2. I assume your outdoor sensor is connected to your thermostat? I'd be curious to hook mine up to my MasterStat. But... that's just logistically not possible, so I won't ;).

**Reply by KO-RAM** [reply]
Yes, I do have my thermocouple connected to my thermostat directly. I just checked register 742 again and it looks like it is the outdoor temperature. Outdoor temperature is 20 deg F on the thermostat and register 742 is: ??? (742): 203 (0x00cb). 

What needs to be done to bring this into home assistant?

**Reply by ccutrer** [reply]
> What needs to be done to bring this into home assistant?

Update to v1.5.7. I switched which register it's sourcing from.

**Reply by KO-RAM** [reply]
That's fast. Thanks, I'll give it a try after work.

**Reply by KO-RAM** [reply]
I updated and everything is working correctly. I now have Outdoor Temperature showing up in Home Assistant for my WaterFurnace. Thanks, this is great way to get a lot of information from the 7-series WaterFurnace equipment, and it keeps it all local and off the internet.
